### PR TITLE
misc: move irdl_definition to IRDLOperation

### LINKED ISSUE #746: Solve circular dependency in ir.py
Currently, `Operation` and `Attribute` have circular dependencies with IRDL through `OpDef` and `ParameterAttrDef`
The recent refactor might allow to solve them now.

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

**Review by PapyChacal — APPROVED** [review]
Nice! I would just argue this does not close the issue itself, as we still have the circular dependency on the attributes side?

**Review by PapyChacal — COMMENTED** [review]
Mh, actually, maybe we need to think about it.
This changes the order of the verification. Specifically here, the Operation is not "ErasedSSAValue-checked" before calling the op's verify.
We might want to go for some kind of hooking here rather than using inheritance? Or any way that would let the base Operation be in control of this order, I mean

**Comment by superlopuh** [comment]
Ah you're right about the circular dependency. I think that the constraint verification makes sense to do before the ErasedSSA check, so no need to add hooks, although I might be missing something.

**Comment by math-fehr** [comment]
I think we need to be careful with that.
We should rather have hooks as Emilien suggested, or calling other functions.
But right now, this changes the order of operations in a non-nice way.

The order should be:
* Checking for ErasedSSAValue
* IRDL verifier
* user custom verifier
* verifying regions
* user custom region verifier, this is something that we don't currently have

I have an idea of design we could talk to tomorrow if you want, but Emilien also has a design idea with hooks that could work!

**Comment by superlopuh** [comment]
Could this be done as follow-up work? I don't see an issue with verifying the erased SSA values after the irdl verify for the time being.

**Comment by tobiasgrosser** [comment]
I would be careful here. Afaiu mlir also vérifies the ssa structure first before going to verify operations. I feel being inconsistent here as a potential of becoming quite costly and confusing in the future.

**Comment by webmiche** [comment]
Thinking again about splitting IRDL and the data structures themselves, shouldn't the verification of the IRDL properties actually be part of a custom verifier? To the data structures, that's essentially what it is, right?

> The order should be:
> 
>     * Checking for ErasedSSAValue
> 
>     * IRDL verifier
> 
>     * user custom verifier
> 
>     * verifying regions
> 
>     * user custom region verifier, this is something that we don't currently have

Where does this order come from? MLIR? Or just intuition? It seems to me that for the operation, both `IRDL Verifier` and `user custom verifier` are custom verifiers built on top of additional semantic that you want to enforce.

**Comment by superlopuh** [comment]
I added back the hacky override in `irdl_op_def`, would be good to migrate to something a little more object-oriented, but at least the PR no longer changes behaviour

File: xdsl/irdl.py [hunk 1]
```diff
@@ -1289,6 +1282,14 @@ def irdl_definition(cls: type[_OpT]):
 
     new_attrs["irdl_definition"] = irdl_definition
 
+    custom_verify = getattr(cls, 'verify_')
+
+    def verify_(self: _OpT):
+        op_def.verify(self)
+        custom_verify(self)
```
Comment (PapyChacal) [inline xdsl/irdl.py:1289]: I think you could just make this a method of IRDLOperation now, instead of generating it in irdl_op_definition?

**Comment by PapyChacal** [comment]
What I'm after is just having this verification order well defined and structured in one place, so it doesn't become a huge code factorization problem in itself later on.
I don't think we have a case or test where the ErasedSSAValue could be an issue yet, though I think we will probably face cornercases with this order. It is all about assumptions that can be made by each abstraction layer, trusting the infrastructure to enforce them as they expect before running their code; assumptions tend to go all over the place, so I feel having a clear base to work on that is very important, to not build tons of workarounds on top of something vague, that will slowly make it harder to simplify.

I also don't find the current override done by IRDL so hacky; It's just wrapping the user-provided verifier in what IRDL provides.

> would be good to migrate to something a little more object-oriented

I'm not sure, in which way and why?

File: xdsl/ir.py [hunk 2]
```diff
@@ -683,11 +683,6 @@ def verify(self, verify_nested_ops: bool = True) -> None:
             if isinstance(operand, ErasedSSAValue):
                 raise Exception("Erased SSA value is used by the operation")
 
-        # Verifier generated by irdl_op_def
-        op_def = type(self).irdl_definition
-        if op_def is not None:
-            op_def.verify(self)
```
Comment (PapyChacal) [inline xdsl/ir.py:689]: So you're not calling IRDL's verify anymore now, right?
I think you could just call `verify_` here for now?

**Review by georgebisbas — APPROVED** [review]

**Review by math-fehr — APPROVED** [review]
:+1:

**Comment by math-fehr** [comment]
> Thinking again about splitting IRDL and the data structures themselves, shouldn't the verification of the IRDL properties actually be part of a custom verifier? To the data structures, that's essentially what it is, right?
> 
> > The order should be:
> > ```
> > * Checking for ErasedSSAValue
> > 
> > * IRDL verifier
> > 
> > * user custom verifier
> > 
> > * verifying regions
> > 
> > * user custom region verifier, this is something that we don't currently have
> > ```
> 
> Where does this order come from? MLIR? Or just intuition? It seems to me that for the operation, both `IRDL Verifier` and `user custom verifier` are custom verifiers built on top of additional semantic that you want to enforce.

Yes, these are MLIR semantics!
We'll make sure to follow these semantics in the next few PRs.

**Pull request merged**